components/camel-asn1/src/main/java/org/apache/camel/dataformat/asn1/ASN1DataFormat.java

```diff
@@ -0,0 +1,177 @@
+
+public class ASN1DataFormat extends ServiceSupport implements DataFormat, DataFormatName {
+    private boolean usingIterator;
+    private Class<?> clazz;
+    private String clazzName;
+
+    public ASN1DataFormat() {
+        super();
+        this.usingIterator = false;
+    }
+
+    public ASN1DataFormat(String clazzName) {
+        super();
+        this.usingIterator = true;
+        this.clazzName = clazzName;
+    }
+
+    public ASN1DataFormat(Class<?> clazz) {
+        super();
+        this.usingIterator = true;
+        this.clazz = clazz;
+    }
+
+    @Override
+    public String getDataFormatName() {
+        return "asn1";
+    }
+
+    @Override
+    public void marshal(Exchange exchange, Object graph, OutputStream stream) throws Exception {
+        InputStream berOut = null;
+        if (usingIterator) {
+            if (clazzName != null || clazz != null) {
+                if (clazz != null && clazz.getName() != null) {
+                    clazzName = clazz.getName();
+                }
+                encodeGenericTypeObject(exchange, clazz, stream);
+                return;
+            }
+            Object record = exchange.getIn().getBody();
+            if (record instanceof ASN1Primitive) {
+                ASN1Primitive asn1Primitive = ObjectHelper.cast(ASN1Primitive.class, record);
+                berOut = new ByteArrayInputStream(asn1Primitive.getEncoded());
+            } else if (record instanceof byte[]) {
+                berOut = new ByteArrayInputStream(ObjectHelper.cast(byte[].class, record));
+            }
+        } else {
+            byte[] byteInput = exchange.getContext().getTypeConverter().mandatoryConvertTo(byte[].class, exchange, graph);
+            berOut = new ByteArrayInputStream(byteInput);
+        }
+        try {
+            IOHelper.copy(berOut, stream);
+        } finally {
+            IOHelper.close(berOut, stream);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void encodeGenericTypeObject(Exchange exchange, Class<?> clazz, OutputStream stream) 
+        throws NoSuchMethodException, SecurityException, 
+        IllegalAccessException, IllegalArgumentException, 
+        InvocationTargetException, IOException {
+        Class[] paramOut = new Class[1]; 
+        paramOut[0] = BerByteArrayOutputStream.class;
+        BerByteArrayOutputStream berOut = new BerByteArrayOutputStream(IOHelper.DEFAULT_BUFFER_SIZE / 256, true);
+        Method encodeMethod = exchange.getIn().getBody().getClass().getDeclaredMethod("encode", paramOut);
+        encodeMethod.invoke(exchange.getIn().getBody(), berOut);
+        stream.write(berOut.getArray());
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public Object unmarshal(Exchange exchange, InputStream stream) throws Exception {
+        if (usingIterator) {
+            if (clazzName != null || clazz != null) {
+                if (clazz != null && clazz.getName() != null) {
```

Comment:
I wonder if it could be a good idea to factorize this redudant code in getClassName() or constructor.